roles/config-pip-packages/tasks/main.yml

```diff
@@ -0,0 +1,19 @@
+---
+
+- name: Install Python software packages
+  pip:
+    name: "{{ item }}"
+  loop: "{{ pip.packages | default([]) }}"
+  when:
+    - pip is defined
+    - pip.virtualenv is not defined
+
+- name: Install Python software packages using {{ pip.virtualenv }}
+  pip:
+    name: "{{ item }}"
+    virtualenv: "{{ pip.virtualenv }}"
+    virtualenv_python: "{{ pip.virtualenv_python | default(None) }}"
+  loop: "{{ pip.packages | default([]) }}"
+  when:
+    - pip is defined
```

Comment:
I know Ansible has some quirks re: how it evaluates checks and at times fail even if a `when` check it done for `is defined`. However, for `loop` I believe the check is honored and this particular code should work as-is. No?

Reply:
The value of `loop` is evaluated before `when`. So this code will throw an error when it attempts to evaluate `pip.packages` because the undefined `pip` will have no property `packages`. The default filter won't prevent the error here. The only case the `when` is evaluated is after `pip` must have been defined.

Reply:
@jkupferer yes, I know Ansible does that in some cases, but don't think that's the case for `loop`. Here's a little playbook to test with - it shows that it skips the loop correctly:
```
- hosts: localhost
  tasks:
  - debug:
      msg: "{{ item }}"
    loop: "{{ pip.packages | default([]) }}"
    when:
      - pip is defined

```

.. at least it does the "right thing" in Ansible 2.8

Reply:
@jkupferer my testing also shows no issues with `pip` being undefined.
I've tested with `2.6.5` and `2.7.7`

Reply:
Ah, cool. So `default` does catch that, that's good to know.
Still... the `pip is defined` check still does nothing because the loop is never executed and the `when` is never evaluated if `pip` is not defined. For example, with `pip` undefined this also runs without error:

```
- hosts: localhost
  tasks:
  - debug:
      msg: "{{ item }}"
    loop: "{{ pip.packages | default([]) }}"
    when:
    - some_undefined_variable
```